src/msw/webview_edge.cpp

```diff
@@ -772,6 +837,65 @@ bool wxWebViewEdge::RunScript(const wxString& javascript, wxString* output) cons
     return true;
 }
 
+bool wxWebViewEdge::AddScriptMessageHandler(const wxString& name)
+{
+    // Edge only supports a single message handler
+    if (!m_impl->m_scriptMsgHandlerName.empty())
+        return false;
+
+    m_impl->m_scriptMsgHandlerName = name;
+    m_impl->UpdateWebMessageHandler();
+
+    return true;
+}
+
+bool wxWebViewEdge::RemoveScriptMessageHandler(const wxString& WXUNUSED(name))
+{
+    m_impl->m_scriptMsgHandlerName.clear();
+    m_impl->UpdateWebMessageHandler();
+    return true;
+}
+
+HRESULT wxWebViewEdgeImpl::OnAddScriptToExecuteOnDocumentedCreatedCompleted(HRESULT errorCode, LPCWSTR id)
+{
+    if (SUCCEEDED(errorCode))
+        m_userScriptIds.push_back(id);
+    return S_OK;
+}
+
+bool wxWebViewEdge::AddUserScript(const wxString& javascript,
+    wxWebViewUserScriptInjectionTime injectionTime)
+{
+    // Currently only AT_DOCUMENT_START is supported
+    if (injectionTime != wxWEBVIEW_INJECT_AT_DOCUMENT_START)
+        return false;
+
+    if (m_impl->m_webView)
+    {
+        HRESULT hr = m_impl->m_webView->AddScriptToExecuteOnDocumentCreated(javascript.wc_str(),
+            Callback<ICoreWebView2AddScriptToExecuteOnDocumentCreatedCompletedHandler>(m_impl,
+            &wxWebViewEdgeImpl::OnAddScriptToExecuteOnDocumentedCreatedCompleted).Get());
+        if (FAILED(hr))
+            return false;
+    }
+    else
+        m_impl->m_pendingUserScripts.push_back(javascript);
+
+    return true;
+}
+
+void wxWebViewEdge::RemoveAllUserScripts()
+{
+    m_impl->m_pendingUserScripts.clear();
+    for (wxVector<wxString>::iterator it = m_impl->m_userScriptIds.begin();
+        it != m_impl->m_userScriptIds.end(); ++it)
+    {
+        HRESULT hr = m_impl->m_webView->RemoveScriptToExecuteOnDocumentCreated(it->wc_str());
+        if (FAILED(hr))
+            wxLogApiError("RemoveScriptToExecuteOnDocumentCreated", hr);
+    }
+}
+
```

Comment:
I am sorry if I am missing something but should we not clear also `m_impl->m_userScriptIds` after the loop?

BTW, AFAICT the code in this file requires C++11: would it not be better to use more readable range-based `for`s for the loops like the one above?

Reply:
You are absolutely right, changed.